### PR TITLE
Ducktype: Adds module isolation by target assembly

### PR DESCRIPTION
**Pull request by tonyredondo**
This `PR` adds Ducktype module isolation by target assembly.

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace/DuckTyping/DuckType.Statics.cs [hunk 1]
```diff
@@ -31,9 +32,51 @@ public static partial class DuckType
...
+            if (targetType.IsGenericType)
+            {
+                foreach (var type in targetType.GetGenericArguments())
+                {
+                    if (type.Assembly != targetAssembly)
+                    {
+                        var name = new AssemblyName($"DuckTypeGenericTypeAssembly.{targetType.Name}_{++_assemblyCount}");
+                        name.Version = targetAssembly.GetName().Version;
+                        var assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(name, AssemblyBuilderAccess.Run);
+                        return assemblyBuilder.DefineDynamicModule("MainModule");
```
Comment (kevingosse) [inline src/Datadog.Trace/DuckTyping/DuckType.Statics.cs:64]: NIT: Could you extract this logic in a helper method, and use the same below?

File: test/test-applications/regression/DuplicateTypeProxy/Program.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    public class Program
+    {
+        public static async Task Main(string[] args)
+        {
+            await RunAsync(typeof(HttpClient));
+
+            for (int i = 0; i < 5; i++)
+            {
+                var assembly = Assembly.LoadFile(typeof(HttpClient).Assembly.Location);
+                await RunAsync(assembly.GetType("System.Net.Http.HttpClient"));
```
Comment (kevingosse) [inline test/test-applications/regression/DuplicateTypeProxy/Program.cs:18]: A few iterations with Assembly.LoadFile is fine, but can you also use a custom ALC, to be sure we cover all the cases?

**Review by kevingosse — CHANGES_REQUESTED** [review]

File: src/Datadog.Trace/DuckTyping/DuckType.Statics.cs (hunk 1 above)
Comment (tonyredondo) [reply src/Datadog.Trace/DuckTyping/DuckType.Statics.cs]: done in 1e77200

File: test/test-applications/regression/DuplicateTypeProxy/Program.cs (hunk 2 above)
Comment (tonyredondo) [reply test/test-applications/regression/DuplicateTypeProxy/Program.cs:18]: done in 1e77200

File: src/Datadog.Trace/DuckTyping/DuckType.Statics.cs [hunk 3]
```diff
@@ -58,24 +58,26 @@ private static ModuleBuilder GetModuleBuilder(Type targetType)
                 {
                     if (type.Assembly != targetAssembly)
                     {
-                        var name = new AssemblyName($"DuckTypeGenericTypeAssembly.{targetType.Name}_{++_assemblyCount}");
-                        name.Version = targetAssembly.GetName().Version;
-                        var assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(name, AssemblyBuilderAccess.Run);
-                        return assemblyBuilder.DefineDynamicModule("MainModule");
+                        return CreateModuleBuilder($"DuckTypeGenericTypeAssembly.{targetType.Name}_{++_assemblyCount}", targetAssembly);
```
Comment (kevingosse) [inline src/Datadog.Trace/DuckTyping/DuckType.Statics.cs:61]: I would have put the `_assemblyCount` suffix inside of the helper method, but oh well 😄

File: src/Datadog.Trace/DuckTyping/DuckType.Statics.cs [hunk 4]
```diff
@@ -32,7 +32,7 @@ public static partial class DuckType
         private static readonly MethodInfo _methodBuilderGetToken = typeof(MethodBuilder).GetMethod("GetToken", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
 
         [DebuggerBrowsable(DebuggerBrowsableState.Never)]
-        private static readonly Dictionary<Assembly, ModuleBuilder> ActiveBuilders = new Dictionary<Assembly, ModuleBuilder>();
+        private static readonly ConditionalWeakTable<Assembly, ModuleBuilder> ActiveBuilders = new ConditionalWeakTable<Assembly, ModuleBuilder>();
```
Comment (kevingosse) [inline src/Datadog.Trace/DuckTyping/DuckType.Statics.cs:35]: I wonder if the assembly can actually be unloaded when dynamic types are referencing it. We should give it a try at some point

**Review by kevingosse — APPROVED** [review]

File: src/Datadog.Trace/DuckTyping/DuckType.Statics.cs (hunk 3 above)
Comment (tonyredondo) [reply src/Datadog.Trace/DuckTyping/DuckType.Statics.cs]: done in 8bb619b

File: src/Datadog.Trace/DuckTyping/DuckType.Statics.cs (hunk 4 above)
Comment (tonyredondo) [reply src/Datadog.Trace/DuckTyping/DuckType.Statics.cs]: we can make it work, but I'll leave that for another PR and another release, changed to `Dictionary` in 8bb619b

**Review by kevingosse — APPROVED** [review]

**Review by zacharycmontoya — APPROVED** [review]
Looks good! Can you invoke the regression test in our smoke test suite before merging?

**Comment by tonyredondo** [comment]
> Looks good! Can you invoke the regression test in our smoke test suite before merging?

done in 6a131ec

**Pull request merged**